### ISSUE TITLE
llama sharded fails to beam on HSA

### ISSUE THREAD

**Issue by wozeparrot** [tinygrad/tinygrad#4612]
Reproduces with #4576 and `DEBUG=2 BEAM=4 python3 examples/llama3.py --model /raid/weights/LLaMA-3/8B/ --shard 6`

```
Traceback (most recent call last):
  File "/home/tiny/tinygrad/examples/llama3.py", line 198, in <module>
    model(Tensor([prompt], device=device), 0, TEMPERATURE, top_k=TOP_K, top_p=TOP_P, alpha_f=ALPHA_F, alpha_p=ALPHA_P).realize()
  File "/home/tiny/tinygrad/extra/models/llama.py", line 175, in __call__
    return self.forward(tokens, start_pos, temperature, top_k, top_p, alpha_f, alpha_p)
  File "/home/tiny/tinygrad/extra/models/llama.py", line 168, in forward
    return ret.realize()
  File "/home/tiny/tinygrad/tinygrad/tensor.py", line 166, in realize
    run_schedule(*self.schedule_with_vars(*lst))
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 186, in run_schedule
    for ei in lower_schedule(schedule):
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 179, in lower_schedule
    while len(schedule): yield lower_schedule_item(schedule.pop(0))
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 165, in lower_schedule_item
    runner = get_runner(si.outputs[0].device, si.ast)
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 137, in get_runner
    prg: Program = get_linearizer(Device[dname].renderer, ast).to_program()
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 35, in get_linearizer
    timed = sorted([(nm, tk, time_linearizer(tk, rawbufs, allow_test_size=False, clear_l2=True)) for nm, tk in lins], key=lambda x: x[2])
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 35, in <listcomp>
    timed = sorted([(nm, tk, time_linearizer(tk, rawbufs, allow_test_size=False, clear_l2=True)) for nm, tk in lins], key=lambda x: x[2])
  File "/home/tiny/tinygrad/tinygrad/engine/search.py", line 180, in time_linearizer
    tms = _time_program(p, dev.compiler.compile(p.src), var_vals, rawbufs,
  File "/home/tiny/tinygrad/tinygrad/engine/search.py", line 48, in _time_program
    tms.append(cast(float, car(rawbufs, var_vals, wait=True))*factor)
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 81, in __call__
    return self.clprg(*[x._buf for x in rawbufs], **lra, vals=tuple(var_vals[k] for k in self.p.vars), wait=wait)
  File "/home/tiny/tinygrad/tinygrad/runtime/ops_hsa.py", line 80, in __call__
    raise RuntimeError(f"HSAProgram.__call__: incorrect args struct size {ctypes.sizeof(self.args_struct_t)} != {self.kernargs_segment_size}")
RuntimeError: HSAProgram.__call__: incorrect args struct size 64 != 24
```

**Comment by wozeparrot** [comment]
Also fails with `examples/llama.py` running `DEBUG=2 BEAM=4 python3 examples/llama.py --gen 3 --size 8B --shard 6 --prompt "Hello."`

```
Traceback (most recent call last):
  File "/home/tiny/tinygrad/examples/llama.py", line 529, in <module>
    tok_tensor = llama.model(Tensor([toks[start_pos:]], device=device), start_pos, args.temperature)
  File "/home/tiny/tinygrad/extra/models/llama.py", line 175, in __call__
    return self.forward(tokens, start_pos, temperature, top_k, top_p, alpha_f, alpha_p)
  File "/home/tiny/tinygrad/extra/models/llama.py", line 168, in forward
    return ret.realize()
  File "/home/tiny/tinygrad/tinygrad/tensor.py", line 166, in realize
    run_schedule(*self.schedule_with_vars(*lst))
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 186, in run_schedule
    for ei in lower_schedule(schedule):
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 179, in lower_schedule
    while len(schedule): yield lower_schedule_item(schedule.pop(0))
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 165, in lower_schedule_item
    runner = get_runner(si.outputs[0].device, si.ast)
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 137, in get_runner
    prg: Program = get_linearizer(Device[dname].renderer, ast).to_program()
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 35, in get_linearizer
    timed = sorted([(nm, tk, time_linearizer(tk, rawbufs, allow_test_size=False, clear_l2=True)) for nm, tk in lins], key=lambda x: x[2])
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 35, in <listcomp>
    timed = sorted([(nm, tk, time_linearizer(tk, rawbufs, allow_test_size=False, clear_l2=True)) for nm, tk in lins], key=lambda x: x[2])
  File "/home/tiny/tinygrad/tinygrad/engine/search.py", line 180, in time_linearizer
    tms = _time_program(p, dev.compiler.compile(p.src), var_vals, rawbufs,
  File "/home/tiny/tinygrad/tinygrad/engine/search.py", line 48, in _time_program
    tms.append(cast(float, car(rawbufs, var_vals, wait=True))*factor)
  File "/home/tiny/tinygrad/tinygrad/engine/realize.py", line 81, in __call__
    return self.clprg(*[x._buf for x in rawbufs], **lra, vals=tuple(var_vals[k] for k in self.p.vars), wait=wait)
  File "/home/tiny/tinygrad/tinygrad/runtime/ops_hsa.py", line 80, in __call__
    raise RuntimeError(f"HSAProgram.__call__: incorrect args struct size {ctypes.sizeof(self.args_struct_t)} != {self.kernargs_segment_size}")
RuntimeError: HSAProgram.__call__: incorrect args struct size 64 != 32
```

**Comment by nimlgen** [comment]
Number of arguments generated in kernel differs to number of rawbuf passed in. uops removes some globals, but they present in linearizer, so `buf_from_lin` creates more rawbufs.

**Comment by chenyuxyz** [comment]
have a fix in #4621 that applies the global buffer filter in search, and it fixes the llama search.

i cannot trigger the test to fail though. @nimlgen does the arg check only happen with multi-gpu? appreciate any suggestion to unit test this

**Comment by nimlgen** [comment]
The arg check happens in any case. `IGNORE_BEAM_CACHE=1` maybe? And `beam_search` just handles `RuntimeError`, so the test will never fail

**Comment by wozeparrot** [comment]
Fixed with #4621
